File: frontend/src/services/practice.ts
// =============================================================================
// practice.ts — client for Marie's practice drill (PRACTICE v0)
// -----------------------------------------------------------------------------
// Endpoints:
//   GET  /api/cases/:slug/scenarios/:id/practice          → the whole page
//   POST /api/cases/:slug/scenarios/:id/practice/sessions → open a sitting
//   POST /api/practice/answers                            → one answer + the read
//   POST /api/practice/answers/:id/help                   → she opened the drawer
//   POST /api/practice/sessions/:id/end                   → close it, get the sheet
//
// Same idioms as `rehearsal.ts`: `authFetch` (credentials + an AbortController
// timeout), `encodeURIComponent` on every path parameter, `readErrorMessage` to
// surface the backend's `{message}`, and every non-2xx throws.
//
// ## Why the answer call gets a LONGER timeout
//
// It makes an LLM call inside the request. The house rule is 30s normally and 90s
// for synthesis; a one-sentence read is far smaller than a synthesis, but it is
// still a model round trip, and a 30s abort would show Marie "your answer was not
// recorded" for an answer the server was in the middle of recording. 90s.
//
// ## What this file deliberately does NOT do
//
// No client-side judging. The red/green sentence comes from the server or does
// not come at all; this module never composes one, and there is no local
// fallback text anywhere in it. The mockup's `judge()` function was a MOCKUP.
// =============================================================================

import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";
import { readErrorMessage } from "./fetchUtils";
import type { NewSitting } from "./practiceFlow";

/** How long an ordinary practice call may take. The house default. */
export const PRACTICE_TIMEOUT_MS = 30000;

/** How long the answer call may take — it makes a model call inside. */
const READ_TIMEOUT_MS = 90000;

/** One question, with everything its two screens render. */
export type PracticeQuestion = {
  id: string;
  /** `george` or `chuck`. */
  side: string;
  /** True when it braids several barrage rows — a third pill. */
  braid: boolean;
  text: string;
  /** The tactic's NAME, already resolved. `null` withdraws the tag. */
  tactic: string | null;
  /** The "Built from: …" line. `null` renders no source line. */
  receipt: string | null;
  braid_rows: string | null;
  watch_for: string | null;
  /** Present together or not at all. */
  pair_said: string | null;
  pair_admitted: string | null;
  /** `null` renders the stored "no receipt for this one" line. */
  stronger: string | null;
  stronger_lean: string | null;
  /** Marie's one line saying what is wrong with this question. `null` = none. */
  flag_note: string | null;
  /**
   * `cross`, `direct` or `redirect` — what the question DOES.
   *
   * Not the same as `side`: Chuck asks both `direct` and `redirect`, and they
   * are dealt, tagged and judged differently. Nothing on this screen infers one
   * from the other.
   */
  kind: string;
  /** The stable handle the deck file uses (`g1`, `r2`), or `null`. */
  deck_key: string | null;
  /** The `deck_key` of the George question a redirect answers, or `null`. */
  follows_key: string | null;
  /**
   * `Answered on 22 Aug`, ALREADY COMPOSED by the server — or `null` when
   * nobody has answered this question.
   *
   * The ONE status a row carries. `null` renders NOTHING at all, not an empty
   * line: an empty status under a question reads as one that failed to load,
   * which is a different fact from "not answered yet". The stored footnote
   * under the list is what tells a reader which they are looking at.
   *
   * ## ⚑ This is not scoped to the person reading it
   *
   * The page is one page for two people. Chuck opens it to read Marie's
   * answers and to print them, so the answer belongs to the QUESTION rather
   * than to the requester. Its retired predecessor (`status`) was user-scoped,
   * because it reported what SHE did in a sitting.
   */
  answered_on: string | null;
  /**
   * True when this question has been deleted — the mechanism is a hide, so the
   * row is gone from every list while her answers keep pointing at it.
   */
  hidden: boolean;
  /** Who drafted it when nobody has reviewed it (`architect`), or `null`. */
  draft_by: string | null;
};

/** One note, as every panel renders it. */
export type PracticeNote = {
  id: string;
  question_id: string | null;
  answer_id: string | null;
  author: string;
  text: string;
  /** `Tue 18 Aug` — composed, so the browser holds no date format. */
  when: string;
  /** `struck Tue 19 Aug`, or `null` while it stands. Its presence strikes it. */
  struck: string | null;
};

/** What changed since her last sitting, composed. */
export type PracticeChanged = {
  heading: string;
  /** The plain-words list behind the fold. */
  items: string[];
};

/** One thing a new question can be attached to, in the add form's picker. */
export type PracticeAttachOption = {
  source_kind: string;
  source_index: number;
  label: string;
};

/** One of Marie's talking points. */
export type PracticePoint = {
  position: number;
  text: string;
  /** `null` renders the stored named-absence line — never a blank. */
  exhibit: string | null;
};

/**
 * Every string the four screens speak.
 *
 * `Record<string, string>` rather than a named field for each: this mirrors a
 * backend block that is generated from one migration, the page reads it by key,
 * and writing eighty names here would produce eighty chances to typo one into a
 * silent `undefined`. The `wordingOf` helper below is the single reader, and it
 * throws by NAME on a key the payload does not carry.
 */
export type PracticeWording = Record<string, string>;

/** The sitting she walked out of, as the start card offers it back. */
export type OpenSession = {
  session_id: string;
  /** `· today 09:57 · George's side · 1 of 5 answered.` — composed server-side. */
  detail: string;
};

/** Everything the page needs, in one response. */
export type PracticeDeck = {
  scenario_id: string;
  code: string;
  title: string;
  /** EMPTY is a legitimate state: the page says "no practice deck yet". */
  questions: PracticeQuestion[];
  points: PracticePoint[];
  last_session_line: string;
  /**
   * When any question in this deck last changed — the deck's own date.
   *
   * ISO-8601, or `null` on a deck with no questions. Read by the print sheets'
   * "deck as of" line: paper outlives the deck it was taken from, and a reader
   * must be able to tell how stale the sheet in his hand is.
   */
  deck_as_of: string | null;
  /** What the "I'd point to…" picker offers, composed and de-duplicated. */
  receipts: string[];
  /** `null` withdraws the blue resume box entirely. */
  open_session: OpenSession | null;
  /** What the editor's add form may attach a new question to. */
  attach_options: PracticeAttachOption[];
  wording: PracticeWording;
};

/** Her four self-check boxes. */
export type SelfCheck = {
  only_asked: boolean;
  accepted_premise: boolean;
  explained_unasked: boolean;
  guessed: boolean;
};

/**
 * The critique's three parts.
 *
 * ## ⚑ Checked BY EYE against `backend/src/dto/practice.rs::ReadPartsDto`
 *
 * Nothing enforces this boundary — these types are hand-written, not generated,
 * and four fields left the deck payload this week with zero TypeScript errors.
 *
 *   pub call: String          → call: string
 *   pub why: String           → why: string      (empty is LEGITIMATE, never null)
 *   pub pointers: Vec<String> → pointers: string[]  (0–3, never absent)
 *   pub keys: Vec<String>     → keys: string[]      (never absent)
 */
export type ReadParts = {
  call: string;
  why: string;
  pointers: string[];
  keys: string[];
};

/**
 * One citable source, as the critique footnotes it.
 *
 * ## ⚑ Checked BY EYE against `ReadSourceDto`
 *
 *   pub key: String  → key: string
 *   pub text: String → text: string
 *
 * ## ⚑ THESE ARE THE WORDS WE SENT, NEVER WORDS THAT CAME BACK
 *
 * The server builds this from the payload the model was GIVEN. That is the
 * whole value of it: if the read cites S2 for a claim S2 does not support,
 * Marie can see S2 say so. Built from the model's own reply instead, a
 * hallucinated citation would render its own supporting evidence. If anything
 * ever makes this list come from the response, it stops being a safeguard and
 * becomes decoration.
 */
export type ReadSource = {
  key: string;
  text: string;
};

/**
 * What the critique shows about the read.
 *
 * ## ⚑ Checked BY EYE against `AnswerResponse`
 *
 *   pub answer_id: Uuid                    → answer_id: string
 *   pub read_text: Option<String>          → read_text: string | null
 *   pub read_ok: Option<bool>              → read_ok: boolean | null
 *   pub read_parts: Option<ReadPartsDto>   → read_parts: ReadParts | null
 *     ⚑ `skip_serializing_if = "Option::is_none"` — ABSENT, not null, when there
 *       are no parts. The client must treat missing and null alike.
 *   pub read_sources: Vec<ReadSourceDto>   → read_sources: ReadSource[]
 *
 * ## Domain note: `read_text` is a LOSSY PROJECTION of `read_parts`
 *
 * Not a second copy. The server derives it with `compose_read_text`, which
 * deliberately drops `why` and keeps only the first pointer — it was built for a
 * screen that showed one sentence. So the screen renders PARTS IF PRESENT, ELSE
 * TEXT, and never both: showing both would print the call and first pointer
 * twice, and the two cannot be reconciled because one is a summary of the other.
 *
 * ## Domain note: text WITHOUT parts is the COMMON case, not an edge
 *
 * Measured on DEV, 2026-08-23: of 14 stored answers, 12 carry `read_text` and
 * only 2 carry parts — 10 have text and no parts, written before T1 shipped in
 * .404. The fallback is the majority path.
 */
export type AnswerResult = {
  answer_id: string;
  /** `null` → the screen shows "no system read this time". */
  read_text: string | null;
  /** `true` green, `false` red, `null` no read. Three states. */
  read_ok: boolean | null;
  /** The three parts, or `null`/absent on an older answer, an abstain or a failure. */
  read_parts: ReadParts | null;
  /** What the cited keys refer to. Empty when nothing was cited. */
  read_sources: ReadSource[];
};

/** One row of Chuck's sheet — every cell already a word. */
export type PracticeSheetRow = {
  number: number;
  from: string;
  tactic: string;
  question: string;
  answer: string;
  mark: string;
  help_opened: boolean;
  help: string;
  /** What she said she would point to. EMPTY withdraws the line. */
  points_to: string[];
};

/** Chuck's sheet, composed. */
export type PracticeSheet = {
  kicker: string;
  heading: string;
  rows: PracticeSheetRow[];
  /**
   * The deck's flagged questions, already composed server-side into the
   * sentences the sheet prints. EMPTY withdraws the whole block.
   */
  flagged: string[];
  /** The block's heading and its sentence. Both empty when `flagged` is. */
  flagged_heading: string;
  flagged_hint: string;
  /** The deck changes made on the day of this sitting. EMPTY withdraws the block. */
  changes: string[];
  /** That block's heading. Empty when `changes` is. */
  changes_heading: string;
};

/**
 * Read one stored string, or throw naming the key.
 *
 * ## Why this throws instead of returning a fallback
 *
 * There is no literal to fall back to (the wording law, v2 §2b), and a missing
 * key means the store and this build disagree. A `?? ""` here would put a blank
 * button in front of a witness; a `?? "Answer"` would put a sentence in the
 * product that no migration can change. Throwing puts the page's own failure
 * notice on screen with the key in it, which is the only honest option.
 */
export function wordingOf(wording: PracticeWording, key: string): string {
  const value = wording[key];
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(
      `The practice page has no stored wording for "${key}". The backend and ` +
        `this build disagree about the wording store; report it to the site ` +
        `administrator.`,
    );
  }
  return value;
}

/** The deck, the points, the words, and the last-session line — one request. */
export async function fetchPracticeDeck(
  slug: string,
  scenarioId: string,
): Promise<PracticeDeck> {
  const response = await authFetch(
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}/scenarios/` +
      `${encodeURIComponent(scenarioId)}/practice`,
    { timeoutMs: PRACTICE_TIMEOUT_MS },
  );

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to load the practice deck (HTTP ${response.status}${detail}).`,
    );
  }

  const parsed = (await response.json()) as Partial<PracticeDeck>;
  // The shapes the page cannot render without. `questions` may be EMPTY — that
  // is the un-seeded scenario, and a legitimate screen — but it must be an
  // array, and `wording` must be there or every label on all four screens is
  // blank.
  if (
    !Array.isArray(parsed.questions) ||
    !Array.isArray(parsed.points) ||
    !Array.isArray(parsed.receipts) ||
    !Array.isArray(parsed.attach_options) ||
    parsed.wording == null ||
    typeof parsed.last_session_line !== "string"
  ) {
    throw new Error(
      "The practice response is missing questions/points/wording — " +
        "backend/frontend contract mismatch. Report it to the site administrator.",
    );
  }
  return parsed as PracticeDeck;
}

/** Open a session. `who` is `george`, `chuck` or `mixed`. */
export async function startPracticeSession(
  slug: string,
  scenarioId: string,
  sitting: NewSitting,
): Promise<string> {
  const response = await authFetch(
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}/scenarios/` +
      `${encodeURIComponent(scenarioId)}/practice/sessions`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        who: sitting.who,
        queue: sitting.queue,
        count: sitting.count,
        skipped_today: sitting.skippedToday,
      }),
      timeoutMs: PRACTICE_TIMEOUT_MS,
    },
  );

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Could not start the practice session (HTTP ${response.status}${detail}).`,
    );
  }
  const parsed = (await response.json()) as { session_id?: string };
  if (typeof parsed.session_id !== "string") {
    throw new Error("The session response carried no session id.");
  }
  return parsed.session_id;
}

/**
 * Record one answer and return the read.
 *
 * ## Why the mark and the boxes are NOT sent here
 *
 * Both are decided after she has read the reveal, and this call is what produces
 * the reveal. The row opens provisional (no boxes, marked fine) and
 * `closePracticeAnswer` settles it when she moves on — which means her typed
 * answer is already stored if she closes the laptop mid-screen.
 */
/**
 * A refused answer, carrying the STATUS as a field.
 *
 * ## Why a class and not a message the page matches on
 *
 * The screen has to tell one refusal apart from the rest: 409, which means this
 * question is already answered in this sitting and this tab is simply behind
 * (two tabs on one sitting — hotfix §3.13). Every other status is "it did not
 * land, try again".
 *
 * Reading that out of `"… (HTTP 409)"` with a regex would work until somebody
 * edited the sentence, and the sentence is one the store may yet own. The
 * status is a number the server sent; carry the number.
 *
 * ## Rust Learning: this is the `thiserror` enum, in TypeScript's idiom
 *
 * The backend would model this as `#[derive(thiserror::Error)] enum AnswerError
 * { AlreadyAnswered, … }` and match on the variant. TypeScript has no enums
 * worth the name here, so the equivalent is a subclass with a discriminating
 * field and an `instanceof` check — same shape, same reason: the CALLER decides
 * what a particular failure means, and it cannot decide from prose.
 */
export class PracticeAnswerError extends Error {
  readonly status: number;

  constructor(status: number, message: string) {
    super(message);
    this.name = "PracticeAnswerError";
    this.status = status;
  }
}

export async function submitPracticeAnswer(input: {
  sessionId: string;
  questionId: string;
  answerText: string;
  dontRecall: boolean;
  /**
   * The receipts she picked, or `null` when she never opened the control.
   *
   * `null` and `[]` are DIFFERENT and are sent differently: an empty array says
   * she looked at the list and reached for nothing, which is a fact about the
   * answer; `null` says the question of what she would point to never came up.
   */
  pointsTo: string[] | null;
}): Promise<AnswerResult> {
  const response = await authFetch(`${API_BASE_URL}/api/practice/answers`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      session_id: input.sessionId,
      question_id: input.questionId,
      answer_text: input.answerText,
      dont_recall: input.dontRecall,
      points_to: input.pointsTo,
    }),
    timeoutMs: READ_TIMEOUT_MS,
  });

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new PracticeAnswerError(
      response.status,
      `Your answer was not recorded (HTTP ${response.status}${detail}).`,
    );
  }
  const parsed = (await response.json()) as Partial<AnswerResult>;
  if (typeof parsed.answer_id !== "string") {
    throw new Error("The answer response carried no answer id.");
  }
  return {
    answer_id: parsed.answer_id,
    read_text: parsed.read_text ?? null,
    read_ok: parsed.read_ok ?? null,
    // `?? null` and not `?? undefined`: the field is ABSENT on the wire when
    // there are no parts (`skip_serializing_if`), and the screen must not have
    // to know the difference between missing and null.
    read_parts: parsed.read_parts ?? null,
    // An empty ARRAY, never undefined. A critique whose sources came back
    // undefined would render its citation keys with nothing behind them, which
    // is the one shape this list exists to prevent.
    read_sources: parsed.read_sources ?? [],
  };
}

/** Settle one answer when she leaves the reveal: her mark and her four boxes. */
export async function closePracticeAnswer(
  answerId: string,
  mark: "fine" | "repeat",
  selfCheck: SelfCheck,
): Promise<void> {
  const response = await authFetch(
    `${API_BASE_URL}/api/practice/answers/${encodeURIComponent(answerId)}/close`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ mark, self_check: selfCheck }),
      timeoutMs: PRACTICE_TIMEOUT_MS,
    },
  );

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Your mark for that question was not recorded (HTTP ${response.status}${detail}).`,
    );
  }
}

/** She opened the stronger-answer drawer; Chuck's sheet says so. */
export async function markHelpOpened(answerId: string): Promise<void> {
  const response = await authFetch(
    `${API_BASE_URL}/api/practice/answers/${encodeURIComponent(answerId)}/help`,
    { method: "POST", timeoutMs: PRACTICE_TIMEOUT_MS },
  );

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Opening the stronger answer was not recorded (HTTP ${response.status}${detail}).`,
    );
  }
}

/** Close the session and get Chuck's sheet. */
export async function endPracticeSession(
  sessionId: string,
): Promise<PracticeSheet> {
  const response = await authFetch(
    `${API_BASE_URL}/api/practice/sessions/${encodeURIComponent(sessionId)}/end`,
    { method: "POST", timeoutMs: PRACTICE_TIMEOUT_MS },
  );

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `The session could not be closed (HTTP ${response.status}${detail}).`,
    );
  }
  const parsed = (await response.json()) as Partial<PracticeSheet>;
  if (!Array.isArray(parsed.rows) || typeof parsed.heading !== "string") {
    throw new Error(
      "Chuck's sheet came back without its rows or its heading — " +
        "backend/frontend contract mismatch.",
    );
  }
  return parsed as PracticeSheet;
}

File: backend/src/repositories/pipeline_repository/practice.rs
//! The practice drill's three tables, read and written.
//!
//! `practice_questions` (the deck), `practice_sessions` (one sitting) and
//! `practice_answers` (the log Chuck's sheet is rendered from). All three live in
//! `colossus_legal_v2` — see the migration's header for why.
//!
//! ## What this module deliberately cannot do
//!
//! There is no `delete_question`, no `update_question` and no "reseed". A stored
//! question is cited by `practice_answers.question_id` under `ON DELETE RESTRICT`,
//! and Chuck's sheet is the record of what Marie was actually asked. Editing the
//! deck is a page in v1 with its own audit; it is not a function some other
//! caller can reach today by accident.
//!
//! ## Rust Learning: `sqlx::FromRow` and runtime queries
//!
//! The record structs derive `FromRow`, which teaches sqlx how to build one from
//! a result row by matching COLUMN NAMES to field names. Combined with
//! `query_as::<_, Record>(sql)` this gives typed rows without the `query!` macro —
//! which matters here because the macro needs a live database (or a checked-in
//! offline cache) at COMPILE time, and these tables did not exist when the
//! container image that builds this code was last prepared.

use sqlx::PgPool;
use uuid::Uuid;

use super::PipelineRepoError;

/// One question from the deck, with everything its reveal screen renders.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct PracticeQuestionRecord {
    pub id: Uuid,
    pub side: String,
    pub text: String,
    /// TACTIC_DECK_v1 card 1–7. `None` on a Chuck question — which has no tactic,
    /// as opposed to having one called "none".
    pub tactic: Option<i16>,
    pub braid_rows: Option<String>,
    pub source_kind: String,
    pub source_ref: Option<String>,
    pub receipt: Option<String>,
    pub watch_for: Option<String>,
    pub stronger: Option<String>,
    pub stronger_lean: Option<String>,
    pub pair_said: Option<String>,
    pub pair_admitted: Option<String>,
    pub sort_order: i32,
    /// Marie's one line saying what is wrong with this question, or `None`.
    /// Read with the deck because the start screen renders it on the row and
    /// Chuck's sheet lists it at the foot — one read serves both.
    pub flag_note: Option<String>,
}

/// One of Marie's three talking points, read live from the scenario record.
///
/// ## Domain note: why the points are NOT stored on the deck
///
/// They are hers, they already exist, and they are edited on the rehearsal page.
/// A copy on the deck would be a second truth that drifts the first time she
/// rewords one — and the wording she practises with has to be the wording she
/// took to Chuck.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct PracticePointRecord {
    /// The number printed beside it: `item_index + 1`.
    pub position: i32,
    pub text: String,
    /// The authored exhibit phrase behind this point, or `None` when nobody has
    /// paired one. `None` renders the stored named-absence line, never a blank.
    pub exhibit: Option<String>,
}

/// A summary of the most recent ENDED session, for the start screen's line.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct LastSessionRecord {
    pub ended_at: chrono::DateTime<chrono::Utc>,
    pub answered: i64,
    pub repeats: i64,
}

/// One row of Chuck's sheet.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct PracticeSheetRow {
    pub side: String,
    pub braid_rows: Option<String>,
    pub tactic: Option<i16>,
    pub question: String,
    pub answer_text: String,
    pub mark: String,
    pub help_opened: bool,
}

/// Everything one answer needs recording. A struct rather than eleven arguments:
/// `insert_answer(pool, a, b, true, false, None, …)` is unreadable at the call
/// site and one transposition away from logging the wrong booleans.
#[derive(Debug, Clone)]
pub struct NewAnswer {
    pub session_id: Uuid,
    pub question_id: Uuid,
    pub answer_text: String,
    pub dont_recall: bool,
    pub read_text: Option<String>,
    pub read_ok: Option<bool>,
    pub read_error: Option<String>,
    pub read_input_tokens: Option<i32>,
    pub read_output_tokens: Option<i32>,
    pub read_ms: Option<i32>,
    pub read_model: Option<String>,
    /// What the model said when this build refused to show it. `None` otherwise.
    pub read_raw_reply: Option<String>,
    pub self_check: serde_json::Value,
    pub mark: String,
}

/// The scenario's deck, in the order it is dealt.
pub async fn list_deck(
    pool: &PgPool,
    scenario_id: Uuid,
) -> Result<Vec<PracticeQuestionRecord>, PipelineRepoError> {
    sqlx::query_as::<_, PracticeQuestionRecord>(
        "SELECT id, side, text, tactic, braid_rows, source_kind, source_ref, receipt, \
                watch_for, stronger, stronger_lean, pair_said, pair_admitted, sort_order, \
                flag_note \
         FROM practice_questions WHERE scenario_id = $1 ORDER BY sort_order",
    )
    .bind(scenario_id)
    .fetch_all(pool)
    .await
    .map_err(PipelineRepoError::from)
}

/// One question by id, for the answer path.
pub async fn get_question(
    pool: &PgPool,
    question_id: Uuid,
) -> Result<Option<PracticeQuestionRecord>, PipelineRepoError> {
    sqlx::query_as::<_, PracticeQuestionRecord>(
        "SELECT id, side, text, tactic, braid_rows, source_kind, source_ref, receipt, \
                watch_for, stronger, stronger_lean, pair_said, pair_admitted, sort_order, \
                flag_note \
         FROM practice_questions WHERE id = $1",
    )
    .bind(question_id)
    .fetch_optional(pool)
    .await
    .map_err(PipelineRepoError::from)
}

/// One seeded stand-in receipt, keyed to the point it backs by printed position.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct PracticePointReceipt {
    pub position: i32,
    pub text: String,
}

/// The seeded receipts for this scenario's points.
///
/// Separate from [`list_points`] rather than joined into it because they are
/// different KINDS of fact: one is the scenario's own record, the other is a
/// stand-in the deck carries until the pairing editor exists. Keeping the two
/// reads apart is what lets the caller state the precedence out loud instead of
/// burying it in a `COALESCE`.
pub async fn list_point_receipts(
    pool: &PgPool,
    scenario_id: Uuid,
) -> Result<Vec<PracticePointReceipt>, PipelineRepoError> {
    sqlx::query_as::<_, PracticePointReceipt>(
        "SELECT position, text FROM practice_point_receipts \
         WHERE scenario_id = $1 ORDER BY position",
    )
    .bind(scenario_id)
    .fetch_all(pool)
    .await
    .map_err(PipelineRepoError::from)
}

/// The scenario's talking points, with the exhibit phrase a human paired.
///
/// `MIN(note)` collapses the m:n link to the one phrase the point shows. A point
/// with two paired exhibits is not a state this surface can render — the
/// rehearsal page has the same shape — and taking the first is the same choice
/// made there rather than a new one invented here.
pub async fn list_points(
    pool: &PgPool,
    scenario_id: Uuid,
) -> Result<Vec<PracticePointRecord>, PipelineRepoError> {
    sqlx::query_as::<_, PracticePointRecord>(
        "SELECT (i.item_index + 1) AS position, i.text, MIN(f.note) AS exhibit \
         FROM scenario_responses r \
         JOIN response_items i ON i.response_id = r.id \
         LEFT JOIN response_item_fact_refs f ON f.response_item_id = i.id \
         WHERE r.scenario_id = $1 \
         GROUP BY i.item_index, i.text \
         ORDER BY i.item_index",
    )
    .bind(scenario_id)
    .fetch_all(pool)
    .await
    .map_err(PipelineRepoError::from)
}

/// The most recently ENDED session for this scenario, and its counts.
///
/// Only ended sessions: a sitting she walked away from is not a session she did,
/// and reporting it as one would put a number on the start screen that says she
/// practised when she did not.
pub async fn last_ended_session(
    pool: &PgPool,
    scenario_id: Uuid,
) -> Result<Option<LastSessionRecord>, PipelineRepoError> {
    sqlx::query_as::<_, LastSessionRecord>(
        "SELECT s.ended_at, \
                COUNT(a.id) AS answered, \
                COUNT(a.id) FILTER (WHERE a.mark = 'repeat') AS repeats \
         FROM practice_sessions s \
         LEFT JOIN practice_answers a ON a.session_id = s.id \
         WHERE s.scenario_id = $1 AND s.ended_at IS NOT NULL \
         GROUP BY s.id, s.ended_at ORDER BY s.ended_at DESC LIMIT 1",
    )
    .bind(scenario_id)
    .fetch_optional(pool)
    .await
    .map_err(PipelineRepoError::from)
}

/// Open a session. Returns the id the answers will cite.
pub async fn start_session(
    pool: &PgPool,
    scenario_id: Uuid,
    who: &str,
    sitting: &NewSitting<'_>,
) -> Result<Uuid, PipelineRepoError> {
    let row: (Uuid,) = sqlx::query_as(
        "INSERT INTO practice_sessions \
         (scenario_id, who, count, queue, skipped_today) \
         VALUES ($1,$2,$3,$4,$5) RETURNING id",
    )
    .bind(scenario_id)
    .bind(who)
    .bind(sitting.count)
    .bind(sitting.queue)
    .bind(sitting.skipped_today)
    .fetch_one(pool)
    .await?;
    Ok(row.0)
}

/// What the sitting was, at the moment it opened.
///
/// A struct rather than three more positional arguments: two of the three are
/// `serde_json::Value` holding arrays of the same shape, and
/// `start_session(pool, id, who, a, b, c)` is one transposition away from
/// recording the questions she kept OUT as the ones she was dealt.
#[derive(Debug, Clone, Copy)]
pub struct NewSitting<'a> {
    /// What she chose off the count pills, or `None` when nothing chose.
    pub count: Option<i32>,
    /// The dealt question ids, in order, as a JSON array.
    pub queue: &'a serde_json::Value,
    /// The ids she kept out on the start screen, as a JSON array.
    pub skipped_today: &'a serde_json::Value,
}

/// The scenario a session belongs to, or `None` if there is no such session.
pub async fn session_scenario(
    pool: &PgPool,
    session_id: Uuid,
) -> Result<Option<Uuid>, PipelineRepoError> {
    let row: Option<(Uuid,)> =
        sqlx::query_as("SELECT scenario_id FROM practice_sessions WHERE id = $1")
            .bind(session_id)
            .fetch_optional(pool)
            .await?;
    Ok(row.map(|r| r.0))
}

/// Close a session so the start screen can report it. Idempotent: a session
/// already ended keeps its first `ended_at`, because the moment she finished is
/// not something a second request should move.
pub async fn end_session(pool: &PgPool, session_id: Uuid) -> Result<(), PipelineRepoError> {
    sqlx::query("UPDATE practice_sessions SET ended_at = NOW() WHERE id = $1 AND ended_at IS NULL")
        .bind(session_id)
        .execute(pool)
        .await?;
    Ok(())
}

/// Record one answered question. Returns the answer's id, which the drawer's
/// help flag later addresses.
pub async fn insert_answer(pool: &PgPool, answer: &NewAnswer) -> Result<Uuid, PipelineRepoError> {
    let row: (Uuid,) = sqlx::query_as(
        "INSERT INTO practice_answers \
         (session_id, question_id, answer_text, dont_recall, read_text, read_ok, read_error, \
          read_input_tokens, read_output_tokens, read_ms, read_model, read_raw_reply, \
          self_check, mark) \
         VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14) RETURNING id",
    )
    .bind(answer.session_id)
    .bind(answer.question_id)
    .bind(&answer.answer_text)
    .bind(answer.dont_recall)
    .bind(answer.read_text.as_deref())
    .bind(answer.read_ok)
    .bind(answer.read_error.as_deref())
    .bind(answer.read_input_tokens)
    .bind(answer.read_output_tokens)
    .bind(answer.read_ms)
    .bind(answer.read_model.as_deref())
    .bind(answer.read_raw_reply.as_deref())
    .bind(&answer.self_check)
    .bind(&answer.mark)
    .fetch_one(pool)
    .await?;
    Ok(row.0)
}

/// Record that she opened the stronger-answer drawer.
///
/// Returns whether a row was touched, so the route can tell "recorded" from "no
/// such answer" rather than reporting success for a write that hit nothing.
pub async fn mark_help_opened(pool: &PgPool, answer_id: Uuid) -> Result<bool, PipelineRepoError> {
    let done = sqlx::query("UPDATE practice_answers SET help_opened = TRUE WHERE id = $1")
        .bind(answer_id)
        .execute(pool)
        .await?;
    Ok(done.rows_affected() == 1)
}

/// Settle an answer when Marie leaves the reveal: her four boxes, and the mark.
///
/// ## Why the answer is written in TWO steps and not one
///
/// The read has to exist before she can react to it, and her four boxes and her
/// "ask me this one again later" both happen AFTER she has read it. One write
/// would mean either recording her answer before the model saw it (losing the
/// read) or holding her typed answer in the browser until she pressed a button
/// (losing it if she walked away). So the row is created when she answers —
/// which is the moment worth surviving a closed laptop — and settled when she
/// moves on.
///
/// Returns whether a row was touched, so the route can tell "settled" from "no
/// such answer" rather than reporting success for a write that hit nothing.
pub async fn close_answer(
    pool: &PgPool,
    answer_id: Uuid,
    mark: &str,
    self_check: &serde_json::Value,
) -> Result<bool, PipelineRepoError> {
    let done = sqlx::query("UPDATE practice_answers SET mark = $2, self_check = $3 WHERE id = $1")
        .bind(answer_id)
        .bind(mark)
        .bind(self_check)
        .execute(pool)
        .await?;
    Ok(done.rows_affected() == 1)
}

/// The session's answers, in the order she gave them — Chuck's sheet.
pub async fn sheet_rows(
    pool: &PgPool,
    session_id: Uuid,
) -> Result<Vec<PracticeSheetRow>, PipelineRepoError> {
    sqlx::query_as::<_, PracticeSheetRow>(
        "SELECT q.side, q.braid_rows, q.tactic, q.text AS question, \
                a.answer_text, a.mark, a.help_opened \
         FROM practice_answers a JOIN practice_questions q ON q.id = a.question_id \
         WHERE a.session_id = $1 ORDER BY a.answered_at, a.id",
    )
    .bind(session_id)
    .fetch_all(pool)
    .await
    .map_err(PipelineRepoError::from)
}

// The SQL in this file is opaque to the compiler (`sqlx::query`, not `query!`),
// so a column name that no migration creates is neither a build error nor a
// unit-test failure — it is a runtime "column … does not exist" on the first
// real request, which is how 2026-08-18 happened. The sibling module reads this
// file and the migrations off disk and asserts they agree (Rule 21), the same
// guard `scan_runs.rs` has carried since the 2026-07-19 defect of that family.
#[cfg(test)]
#[path = "practice_sql_shape.rs"]
mod sql_shape;

#[cfg(test)]
#[path = "practice_schema_tests.rs"]
mod schema_tests;
